### DISCUSSION TITLE
Setting up Apollo with prisma2 and nexus + Comments about the docs

### DISCUSSION THREAD

**Discussion by sfratini** [prisma/prisma#3077] [Q&A] [answered]
Hello! First of all, let me start by saying that I have a lot of experience in several fields but I am a complete noob in GraphQL. I just started learning this we want to implement this in our AWS migration as it seems the way to go. Specially since I was sold at how easy prisma and nexus can reduce the boilerplate and the development cycles which we desperately need. I understand this takes effort, planning, hard work and lots of hours. This array of tools together with nexus and prisma seems to be a great ecosystem. 

That being said, my progress over the past week has been quite stressful and frustrating. I am not a fan of the current docs for prisma and nexus. While it includes much content, it assumes a lot of things for newcomers like me. It is like saying to someone that wants to drive "here you have the clutch, with it you change gears". Well ok...when do I change gears? why do I want to do that? Is my overall thought process over the entire docs. Perhaps the fact the old docs are still live and the fact that prisma1 and prisma2 are two completely different tools, together with repos changing names, tools changing names and interfaces, have not help me on what is was supposed to be a "quick get started to see how cool this is". I know this sounds like a rant, but I am trying to expose why I had to make this post, because I was not able to understand the steps required to do, in my head, such a simple task which is having prisma, nexus and a popular graphql server running. 

So, where I am right now? I was able to have the prisma2 client running against a postgres database. I can run it through introspect and get the .prisma file. I also installed and used prisma plugin so I can generate my nexus schema. Like this for example:

Note that almost all of the examples I could find online to integrate nexus and prisma were about prisma1, and the only guide to migrate that shed some light were in the prisma docs, while I was searching like crazy in the nexus/plugin docs, since well, it was about the nexus and not prisma. 

```
//graphql.ts

import { schema } from 'nexus'

schema.objectType({
  name: 'categories',
  definition(t){
    t.model.id()
  }
})

export const Query = schema.queryType({
    definition(t) {
      t.crud.categories()
    },
  })
  
  export const Mutation = schema.mutationType({
    definition(t) {
      t.crud.createOnecategories()
    }
  })

```

And this is how I have the plugin setup (BTW, another example, the docs say "you can enable crud which is experimental" but that was not clear what was the benefit of that. I now understand it allow you to export the methods in the definitions so it is easier to generate the nexus schema):

```
import { use, settings} from 'nexus'
import { prisma } from 'nexus-plugin-prisma'

settings.change({
    logger: {
        pretty: true
    },

    server: {
        startMessage: (info) => {
        settings.original.server.startMessage(info)
        },
    },

    schema: {
        generateGraphQLSDLFile: './api/apollo/schema.graphql'
    }

})

use(
    prisma(
        {
            migrations: false,
            features: {
                crud: true
            }
        }
    )
)

```

Now this is my simple apollo server:

```
//server.ts

import { ApolloServer } from 'apollo-server-express';
import { createContext } from './context';
import { createServer } from 'http';
import express from 'express';
import { schema } from '../graphql/schema';

const { PORT = 5000 } = process.env;

const app = express();
const server = createServer(app);
// const apollo = new ApolloServer({ typeDefs, resolvers });
const apollo = new ApolloServer({
  schema,
  context: createContext,
  introspection: process.env.NODE_ENV !== 'production',
  playground: process.env.NODE_ENV !== 'production',
});
apollo.applyMiddleware({ app });

server.listen({ port: PORT }, () => {
  process.stdout.write(
    `🚀 Apollo Server ready at http://localhost:${PORT}${apollo.graphqlPath}\n`,
  ); 
});

```

And my schema.ts which I understand this is just prisma1 and WONT work for prisma2, although please let me know if I am mistaken:

```
//graphql/schema.ts
import { makeSchema } from '@nexus/schema';
import { nexusSchemaPrisma } from 'nexus-plugin-prisma/schema';
import { Query, Mutation } from './graphql';
import * as path from 'path'

export const schema = makeSchema({

    shouldGenerateArtifacts: true,

    types: [
        Query,
        Mutation
    ],

    plugins: [
      nexusSchemaPrisma()
    ],

    typegenAutoConfig: {
        sources: [
            {
                source: '@prisma/client',
                alias: 'prisma'
            },
        ],
    },
    
    outputs: {
        schema: path.join(__dirname, './schema.graphql'),
        typegen: path.join(__dirname, './nexus.ts'),
    },

});
```

Now, do I need this? Is this the right way? I get an error which is this if I run it this way: 
```
SyntaxError: Unexpected token {
```
However I understand this can just be the TS transpiler not liking something. I am still digging through that but I am still not sure if this is caused by an error in my code or just an error in configuration because of the way I generate and pass the schema. 

Now, looking at this example: https://github.com/prisma/prisma-examples/blob/master/typescript/graphql-sdl-first/src/schema.ts I am trying to connect nexus to the apollo server as I write this. 

Any push in the right direction and a note as to understand how everything works together would be great. I still see a lot of overlap and confusion as to how all these tools work together. 

Thanks!

**Comment by sfratini** [comment]
Alright the Unexpected token issues were my fault, I was using the wrong node version. The issue that I am having now is that my schema.graphql and types.ts files are not including my complete schema, they are just empty. 

This is my schema.ts file now:

```
import { makeSchema } from '@nexus/schema';
import { nexusSchemaPrisma } from 'nexus-plugin-prisma/schema';
import { Query, Mutation } from './graphql';
import * as path from 'path'

export const schema = makeSchema({

    shouldGenerateArtifacts: true,

    types: [
        Query,
        Mutation
    ],

    plugins: [
      nexusSchemaPrisma({
          experimentalCRUD: true,
      })
    ],

    typegenAutoConfig: {
        sources: [
            {
                source: '@prisma/client',
                alias: 'prisma'
            },
        ],
    },
    
    outputs: {
        schema: path.join(__dirname, './schema.graphql'),
        typegen: path.join(__dirname, './typegen.ts')
    }

});
```

**Comment by sfratini** [comment]
Some more information, if I run and setup the nexus server like this, the schema generated has everything but I want to use apollo. I don't see where the difference is, except that I am using different imports. 

```
import { use, settings} from 'nexus'
import { prisma } from 'nexus-plugin-prisma'

settings.change({
    logger: {
        pretty: true
    },

    server: {
        startMessage: (info) => {
        settings.original.server.startMessage(info)
        },
    },

    schema: {
        generateGraphQLSDLFile: './api/apollo/schema.graphql'
    }

})

use(
    prisma(
        {
            migrations: false,
            features: {
                crud: true
            }
        }
    )
)
```

**Comment by sfratini** [comment, marked as answer]
Made it work! I answered myself on my previous comment. I was using the "schema" from "nexus" but I needed the proper imports from "@nexus/schema". 

These is the final files that work:

```
//graphql.ts

import { objectType, queryType, mutationType } from '@nexus/schema'

export const Category = objectType({
  name: 'categories',
  definition(t){
    t.model.id()
  }
})

export const Query = queryType({
    definition(t) {
      t.crud.categories()
    },
  })
  
  export const Mutation = mutationType({
    definition(t) {
      t.crud.createOnecategories()
    }
  })
```

```
//schema.ts

import { makeSchema } from '@nexus/schema';
import { nexusSchemaPrisma } from 'nexus-plugin-prisma/schema';
import * as graphqlTypes from './graphql';
import * as path from 'path'

export const schema = makeSchema({

    shouldGenerateArtifacts: true,

    types: graphqlTypes,

    plugins: [
      nexusSchemaPrisma({
          experimentalCRUD: true,
          shouldGenerateArtifacts: true
      })
    ],

    typegenAutoConfig: {
        sources: [
          {
            source: '@prisma/client',
            alias: 'client'
          }
        ]
      },
 
    outputs: {
        schema: path.join(__dirname, './schema.graphql'),
        typegen: path.join(__dirname, './typegen.ts')
    }

});
```

This has been quite a trial/error as like I said, the docs and examples kept sending me across prisma, then nexus schema, then nexus, then prisma1 for some reason. Now, I can finally understand how it works but like I said, it is a lot for newcomers and it is quite easy to abandon the project for things like Hasura where everything "just works" (and I tried that one before prisma/nexus). I saw the benefits from this approach and that is why I kept trying, but not many people will, and that will hurt adoption. I am going to leave this open for a while, in case I can get a response or some people might be interested in the exchange. 
Thanks

**Comment by jasonkuhrt** [comment]
Hey @sfratini Am I correct that you never found https://nexusjs.org/migration-guides/prisma-users along your journey?

It looks like you alternated between using the framework and using the library.

In the end IIUC you settled for using the library. Was this because the framework doesn't support custom servers yet, and you would like to use Apollo Server right now?

**Comment by sfratini** [comment]
@jasonkuhrt  Hey! I did found that guide after a while, but did not provide be with enough information. And that guide is for nexus, and I wanted to use the "@nexus/schema". And by saying "I wanted to use that one" is because I understood it was the "newer version" and "nexus" was just the old one. After a while I understood "nexus" is the framework and "@nexus/schema" is just the schema standalone. I think it would make much more sense to keep everything under @nexus as my brain automatically assign those to be the newer versions as most module are going that route. 

As for using the library yes, I wanted to use Apollo because as for my research nexus was not allowing custom servers and using "graphql-yoga" which some people say it is using an older version of Apollo. 

In the end I understood my mistakes and I liked the automatic refreshed schema that came with the nexus server, but I am not 100% sold on using something that is locked (the server). 

Please don't take my comments the wrong way. I think all these tools are impressive and will save me a lot of headaches but I did felt like jumping between sinking boats for a while. 

As for the nexus vs @nexus/schema I am not sure how would nexus help me more than the schema standlone does right now.

**Reply by janpio** [reply]
Don't worry @sfratini - you did nothing wrong and we are happy to take your experience as input to improve the situation as much as possible 👍

**Comment by jasonkuhrt** [comment]
`graphql-yoga` is not maintained and should not be used for new projects.

There is a tutorial for nexus if you're ever curious https://nexusjs.org/getting-started/tutorial/chapter-introduction.

Support for custom servers in `nexus` like Apollo server will eventually come.

**Reply by sfratini** [reply]
@jasonkuhrt  Yep! That is what I saw somewhere and I guessed I understood nexus was still using graphql-yoga under the hood. 

I have read a bit more about the docs, and the information you sent and I seem to have a better view about the nexus framework and library. 

As a rule of thumb, when it comes to "wrapper" frameworks (as in nexus, offering a ready to go graphql server), I only do it when the boilerplate that I save is great or it saves me manual steps that I would not be able to automate otherwise. 

In this case, at least for my specific use case at this moment nexus "saves" me about one 20 line file, but I get to use a server which I can customize to my liking and that I can update on my own if needed. 

I have (or maybe we all have?) suffered from those projects which suddenly are using an older version of a dependency that is breaking your project so I like that flexibility to have the standalone server.

**Reply by jasonkuhrt** [reply]
> As a rule of thumb, when it comes to "wrapper" frameworks

If the definition is a tool that presets other tools, then `nexus` is not that.

The flexibility to bring your own server isn't something `nexus` will lock you out of. It does right now just because we have a missing feature or two.

> I have (or maybe we all have?) suffered from those projects which suddenly are using an older version of a dependency that is breaking your project so I like that flexibility to have the standalone server.

There are certain dependencies that we bundle intentionally and we think the net benefits outweigh the negatives. These include `graphql`, `typescript`. On Prisma plugin we bundle `@prisma/client` and a few more. If you're curious why we do that you can read our rationale here https://nxs.li/why/bundle-dependencies. 

Mentioned there, but worth reiterating, a developer is not locked in. In an emergency they have access to yarn resolutions to control the version. 

> In this case, at least for my specific use case at this moment nexus "saves" me about one 20 line file

If you're open to it I would be curious to see your project.

**Reply by sfratini** [reply]
> If the definition is a tool that presets other tools, then nexus is not that.

Understood. Yes, the more I read about the docs the more I understand how everything works together. I do see the benefit of having everything as a framework and I might just go with nexus framework as the project advances and we need more features. 

> Mentioned there, but worth reiterating, a developer is not locked in. In an emergency they have access to yarn resolutions to control the version.

I just read this and I guess it makes sense and I agree on the concept and decision. I guess my goal right now is to find that sweet spot between the entire abstraction that you get from tools like Firebase where you get zero control but it is very easy to get started and AWS, where you have full flexibility but setting up everything is just very tedious. Even with Amplify and AppSync the setup becomes a mess. Also, you end up with huge cloudformation templates that generate resources with no standard naming convention so after a while you don't even understand which resources are meant to what. So in that sense, I believe nexus/prisma fit right where I want right now. 

> If you're open to it I would be curious to see your project. 

Yeah, I am actually working on two projects where we intend to use nexus/prisma/apollo. 

1. The first one for the company that I work for right now. We have our infrastructure in a mix of Firebase hosting, mongoDB, angularJS, react native, endpoints in a propietary tool made in NodeJS, and a propietary user pool and other things which makes the devops life a living hell. So the goal there is to go to AWS. We have about 180 REST endpoints which we do not want to migrate manually. At the same time, our data access patterns are random, but the data itself is highly relational, features change constantly, we also allow custom development and the code base is badly designed. So, bottom line, NoSQL and REST was probably not the best choice. I inherit this and this is what I want to change
Goal on this project is to move to AWS on PostgreSQL + GraphQL + SQS + API Gateway + AWS Cognito (Auth is the next step I want to deal with and I'll take a look at graphql-shield). 
2. The other one is a personal project where I am building an app/community which I cannot explain much over this forum but basically same concept. I might have an app/web that will change a lot until it gets mature so I am looking for flexibility but also performance. I love firebase for example in the sense that you get neat features like subscription, "cached" data (I refuse to call it offline features as it is now 100% offline), and a nice abstraction over Firestore (which in this case is replaced by prisma model). Problem about firestore is the query system is very limited as do the security rules. No native geospatial support so you need to do a few hacks but waste bandwidth and read requests (so more money that you pay for nothing), limited sorting, limited fields, etc. 

Like saying before my next steps in this process is to check:
- Authentication/Authorization 
- How to measure/monitor performance 
- How to protect from DDOS/people requesting everything from the database. 

I'd be happy to discuss more about the projects or anything in general just can't cover much more publicly. 
Thanks!

**Comment by jk21** [comment]
I found this really helpful. Was trying to setup the same things and the docs are really not clear about it..

https://medium.com/dooboolab/how-to-set-up-a-graphql-server-with-apollo-server-prisma2-and-nexus-schema-f59aa23b5d5d

**Comment by macrozone** [comment]
nexus framework comes with its own server, so you don't need to setup apollo server. It uses apollo server under the hood in the latest version.

You can also use it in a nextjs project, check out the examples!

**Reply by sfratini** [reply]
> nexus framework comes with its own server, so you don't need to setup apollo server. It uses apollo server under the hood in the latest version.
> 
> You can also use it in a nextjs project, check out the examples!

I understand that, but I wanted to use another server. I ended up using the latest version of nexus instead of nexus schema because I did not want to spend more time on it, but still that does not mean that this feature should not work if it is available.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
